Print with Print server (CUPS)

mhbu50 opened the pull request:
Allow printing to a network printer

mhbu50 committed:
solved

mhbu50 committed:
enable print server

mhbu50 committed:
Merge branch 'print_server' of https://github.com/mhbu50/frappe into print_server

mhbu50 committed:
Revert "solved"

This reverts commit 9d57661c445fb8d5cc407d76ae28f44b9c4b05e1.

mhbu50 committed:
set defult

mhbu50 committed:
Update print_settings.py

rmehta commented:
@mhbu50 👍 

Can you add some animated GIF and update docs?

https://github.com/frappe/erpnext/wiki/Pull-Request-Guidelines

mhbu50 committed:
Update print_format.py

mhbu50 commented:
@rmehta sure, I will do that
I need some help, to install CUPS it needs to install library:

`sudo apt-get install libcups2-dev`

where I have to write this command in this PR?

mhbu50 commented:
@nabinhait  If I not install this library before update It will give this error:

`Command "/home/mohd/new-bench/env/bin/python -u -c "import setuptools, tokenize;__file__='/tmp/pip-build-GkeRZA/pycups/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /tmp/pip-hH4Axj-record/install-record.txt --single-version-externally-managed --compile --install-headers /home/mohd/new-bench/env/include/site/python2.7/pycups" failed with error code 1 in /tmp/pip-build-GkeRZA/pycups/
Traceback (most recent call last):
  File "/usr/local/bin/bench", line 11, in <module>
    load_entry_point('bench', 'console_scripts', 'bench')()
  File "/home/mohd/.bench/bench/cli.py", line 40, in cli
    bench_command()
  File "/usr/local/lib/python2.7/dist-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/usr/local/lib/python2.7/dist-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/lib/python2.7/dist-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/lib/python2.7/dist-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/home/mohd/.bench/bench/commands/update.py", line 60, in update
    _update(pull, patch, build, bench, auto, restart_supervisor, restart_systemd, requirements, no_backup, force=force, reset=reset)
  File "/home/mohd/.bench/bench/commands/update.py", line 76, in _update
    update_requirements(bench_path=bench_path)
  File "/home/mohd/.bench/bench/utils.py", line 435, in update_requirements
    install_requirements(pip, req_file)
  File "/home/mohd/.bench/bench/utils.py", line 499, in install_requirements
    exec_cmd("{pip} install -q -r {req_file}".format(pip=pip, req_file=req_file))
  File "/home/mohd/.bench/bench/utils.py", line 159, in exec_cmd
    raise CommandFailedError(cmd)
bench.utils.CommandFailedError: ./env/bin/pip install -q -r ./apps/frappe/requirements.txt
`

mhbu50 committed:
Add frm.trigger for server_ip & port

mhbu50 commented:
animated GIF : 

![anim](https://user-images.githubusercontent.com/11526423/42140991-51320dae-7dae-11e8-9780-ae3f88515bde.gif)

Doc : 

https://github.com/erpnext/foundation/pull/227

strixaluco commented:
I think you should add CUPS dependency to `frappe/requirements.txt`

mhbu50 commented:
@strixaluco it is already there

strixaluco commented:
True, missed that. Then should be something else.

mhbu50 commented:
@nabinhait @codingCoffee  I need to install this library 

`sudo apt-get install libcups2-dev`
where I have to write this command in this PR?

rmehta commented:
You need to add it to bench!

mhbu50 commented:
The issue has been solved by : 

https://github.com/frappe/bench/pull/699

https://github.com/frappe/bench/pull/700

And get merged

Please merge this PR

bghayad commented:
Why to close it?
We are waiting this feature since long time to be added. Please let us push to merge it.
@mhbu50, thanks for the great work. Do you have any idea if an restaurant table will be added for POS? Anyone is working on it?
Regards
Bilal

rmehta commented:
@codingCoffee can you test and approve?

codingCoffee committed:
minor fixes

- [x] linting fixes
- [x] removed format as a parameter, instead used print_format
- [x] removed pycups as a requirement, user needs to manually install
- [x] moved import from global to inside a function to prevent errors
- [x] uncheck Enable Print Server is pycups is not installed

codingCoffee commented:
You should always look at codacy errors!

- [x] linting fixes
- [x] variables like format shouldn't be used in python, since they are built-in variables

Need a patch to fix existing systems by installing cups server, otherwise `pip install cups` shall fail

- [ ] patch to apt / yum install cups on existing systems

Utility

- @mhbu50 correct me if I'm wrong but this is only useful if you have an in house server hosting ERPNext, and your CUPS server and your ERPNext instance are on the same network? Is there any other scenario it is supposed to cover?

codingCoffee committed:
Merge branch 'develop' into print_server

codingCoffee commented:
This PR is now mergeable!

- [x] linting fixes
- [x] removed format as a parameter, instead used print_format
- [x] removed pycups as a requirement, user needs to manually install
- [x] moved import from global to inside a function to prevent errors
- [x] uncheck Enable Print Server is pycups is not installed

@mhbu50 you need to:

- [ ] modify the docs to provide instructions to manually install cups on debian, ubuntu and centos. And to install pycups in the env using the command `./env/bin/pip install pycups`, executed from the `frappe-bench` directory.

We cannot add `pycups` in the the `requirements.txt` since it'll break in case cups is not installed on the server, and secondly it'll break existing installations, and there is not way to install it via patches, since it needs sudo access to install it.

Please do needful, and I shall merge the PR.

mhbu50 commented:
@codingCoffee Thank you for your time to review my PR
I think I solve the issue of installation pycups in these PRs:

https://github.com/frappe/bench/pull/699
https://github.com/frappe/bench/pull/700

If it is not, please let me know to do the needful

codingCoffee commented:
frappe/bench#699, frappe/bench#700 these PRs will make sure cups is installed on new installations, however, there's no way to install them on existing installations using patches. The users will have to manually install CUPS on existing installations. 

Please update the documentation to mention steps for manual installation, based on OS, and finally this command `./env/bin/pip install pycups` in their `frappe-bench` folder

mhbu50 commented:
@codingCoffee  I just update the doc https://github.com/erpnext/foundation/pull/276

zkhan44499 commented:
@mhbu50  Hi, How can we use this feature?
